modules/join/src/main/java/org/apache/ignite/internal/join/ClusterManagementGroupManager.java

```diff
@@ -0,0 +1,198 @@
+
+    /** Prevents double stopping the component. */
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    private final ClusterService clusterService;
+
+    private final Loza raftManager;
+
+    private final RestModule restModule;
+
+    /** Handles cluster initialization flow. */
+    private final ClusterInitializer clusterInitializer;
+
+    /** Constructor. */
+    public ClusterManagementGroupManager(ClusterService clusterService, Loza raftManager, RestModule restModule) {
+        this.clusterService = clusterService;
+        this.raftManager = raftManager;
+        this.restModule = restModule;
+        this.clusterInitializer = new ClusterInitializer(clusterService);
+
+        MessagingService messagingService = clusterService.messagingService();
+
+        var msgFactory = new InitMessagesFactory();
+
+        messagingService.addMessageHandler(InitMessageGroup.class, (msg, addr, correlationId) -> {
+            if (!busyLock.enterBusy()) {
+                if (correlationId != null) {
+                    messagingService.respond(addr, errorResponse(msgFactory, new NodeStoppingException()), correlationId);
+                }
+
+                return;
+            }
+
+            try {
+                if (msg instanceof CancelInitMessage) {
+                    log.info("CMG initialization cancelled, reason: " + ((CancelInitMessage) msg).reason());
+
+                    raftManager.stopRaftGroup(CMG_RAFT_GROUP_NAME);
+
+                    // TODO: drop the Raft storage as well, https://issues.apache.org/jira/browse/IGNITE-16471
+                } else if (msg instanceof CmgInitMessage) {
+                    assert correlationId != null;
+
+                    String[] nodeIds = ((CmgInitMessage) msg).cmgNodes();
+
+                    List<ClusterNode> nodes = resolveNodes(clusterService, Arrays.asList(nodeIds));
+
+                    raftManager.prepareRaftGroup(CMG_RAFT_GROUP_NAME, nodes, CmgRaftGroupListener::new)
+                            .whenComplete((service, e) -> {
+                                if (e == null) {
+                                    Peer leader = service.leader();
+
+                                    assert leader != null;
+
+                                    messagingService.respond(addr, leaderElectedResponse(msgFactory, leader), correlationId);
+                                } else {
+                                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                                }
+                            });
+                }
+            } catch (Exception e) {
+                log.error("Exception when initializing the CMG", e);
+
+                if (correlationId != null) {
+                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                }
+            } finally {
+                busyLock.leaveBusy();
```

Comment:
`busyLock` is needed to prevent the asynchronous operations from starting if the node is stopped. Otherwise we will obtain the necessary futures and will wait for them inside the `stop` method

Reply:
It turns out this is only needed to forbid any method execution when the component is stopping. The code does not communicate this intention too clearly, so it makes sense to improve this part (in all components, as it is done similarly there). This should not be changed in this PR, we'll discuss with the team first and, possible, a new ticket will be filed for this refactoring.